### PR TITLE
Fixes Issue 19212 - Add versions for C++ runtimes.

### PR DESCRIPTION
**Pull request by TurkeyMan**
Necessary for `std::` stuff

Spec: https://github.com/dlang/dlang.org/pull/2461

### CODE REVIEW

**Comment by thewilsonator** [comment]
This will presumably need a spec PR to reserve those versions, and add them to the test that check reserved versions.

**Comment by TurkeyMan** [comment]
Spec where?
I didn't find any such test...

**Comment by thewilsonator** [comment]
`test/fail_compilation/reserved_version.d` and `test/fail_compilation/reserved_version_switch.d`

**Comment by TurkeyMan** [comment]
When you say 'spec', do you mean this: https://github.com/dlang/dlang.org/pull/2461?

**Comment by jpf91** [comment]
Spec: https://dlang.org/spec/version.html#predefined-versions

**Comment by TurkeyMan** [comment]
That was with the initial PR... it's references at the top.

**Comment by TurkeyMan** [comment]
Errr... what's with the CI? O_o

**Comment by jpf91** [comment]
> That was with the initial PR... it's referenced at the top.

Oh right, I didn't see that. Sorry!

File: test/fail_compilation/reserved_version.d [hunk 1]
```diff
@@ -83,6 +83,10 @@ fail_compilation/reserved_version.d(182): Error: version identifier `CRuntime_Gl
 fail_compilation/reserved_version.d(183): Error: version identifier `CRuntime_Microsoft` is reserved and cannot be set
 fail_compilation/reserved_version.d(184): Error: version identifier `CRuntime_Musl` is reserved and cannot be set
 fail_compilation/reserved_version.d(185): Error: version identifier `CRuntime_UClibc` is reserved and cannot be set
+fail_compilation/reserved_version.d(185): Error: version identifier `CppRuntime_DigitalMars` is reserved and cannot be set
+fail_compilation/reserved_version.d(185): Error: version identifier `CppRuntime_LibCpp` is reserved and cannot be set
+fail_compilation/reserved_version.d(185): Error: version identifier `CppRuntime_Microsoft` is reserved and cannot be set
+fail_compilation/reserved_version.d(185): Error: version identifier `CppRuntime_StdCpp` is reserved and cannot be set
```
[the remark this answers is not in the document]
Comment (TurkeyMan) [reply test/fail_compilation/reserved_version.d]: Oh wow, solame.

File: src/dmd/cond.d [hunk 2]
```diff
@@ -653,6 +653,10 @@ extern (C++) final class VersionCondition : DVCondition
             case "CRuntime_Microsoft":
             case "CRuntime_Musl":
             case "CRuntime_UClibc":
+            case "CppRuntime_DigitalMars":
+            case "CppRuntime_LibCpp":
```
Comment (jacob-carlborg) [inline src/dmd/cond.d]: This is never used as far as I can see.

**Review by jacob-carlborg — CHANGES_REQUESTED** [review]

File: src/dmd/cond.d (hunk 2 above)
Comment (jacob-carlborg) [reply src/dmd/cond.d]: > OSX? Clang distributes libc++ as standard no?

It's used on macOS yes.

> Oh, yeah.. Does DMD have any builds that build against such a C++ toolset?

Yes, on macOS.

Comment (jacob-carlborg) [reply src/dmd/cond.d]: But for the C standard library on macOS there is no version identifier. In the code `OSX` is used.

Comment (TurkeyMan) [reply src/dmd/cond.d]: Why isn't there some lines configuring DMD for OSX right there next to the lines I edited?
What should the proper patch for OSX look like? Can you add a patch for OSX?

Comment (TurkeyMan) [reply src/dmd/cond.d]: Added for OSX + FreeBSD, they seem to be the 2 platforms that use it.

File: src/dmd/mars.d [hunk 3]
```diff
@@ -1361,13 +1361,25 @@ void addDefaultVersionIdentifiers()
...
     }
     else static if (TARGET.Linux)
     {
         VersionCondition.addPredefinedGlobalIdent("CRuntime_Glibc");
+        VersionCondition.addPredefinedGlobalIdent("CppRuntime_StdCpp");
+    }
+    else static if (TARGET.OSX || TARGET.FreeBSD)
+    {
+        // why u no CRuntime guys?
+        VersionCondition.addPredefinedGlobalIdent("CppRuntime_LibCpp");
```
Comment (TurkeyMan) [inline src/dmd/mars.d]: As far as I could tell, these are the only guys that use clang as default toolchain...

Comment (TurkeyMan) [reply src/dmd/mars.d]: @jacob-carlborg Does this look correct to you?

Comment (TurkeyMan) [reply src/dmd/mars.d]: Probably not on macos, but certainly on linux, and even windows (mingw/cygwin). Toolchains could be built to target all sorts of combinations. LDC/GDC need the freedom to define these however the platform desires.

Comment (jacob-carlborg) [reply src/dmd/mars.d]: I'm specifically talking about macOS here. I'll let someone else decide if this is necessary or not.

Comment (TurkeyMan) [reply src/dmd/mars.d]: I don't want to be writing:
```D
version(CppRuntime_LibCpp)
  version = CppRuntime_LibCpp_or_OSX;
else version (OSX)
  version = CppRuntime_LibCpp_or_OSX;
version (CppRuntime_LibCpp_or_OSX)
{
  // libc++ code...
}
```

Comment (jacob-carlborg) [reply src/dmd/mars.d]: You would be writing:

```d
version (CppRuntime_StdCpp)
else version (OSX)
else version (FreeBSD)
```

That's how it's done for the C standard library: https://github.com/dlang/druntime/blob/master/src/core/stdc/locale.d#L60-L137.

Comment (ibuclaw) [reply src/dmd/mars.d]: What about using `CppRuntime_GCC` and `CppRuntime_Clang`?

If I were to review code with `StdCpp` and `LibCpp` I would certainly have a hard time remembering the difference.

Comment (ibuclaw) [reply src/dmd/mars.d]: > I'm specifically talking about macOS here.

C++ STL is tied to the compiler toolchain, not the OS.

Comment (jacob-carlborg) [reply src/dmd/mars.d]: Fine. What about the rest of the platforms?

Comment (TurkeyMan) [reply src/dmd/mars.d]: Rest of the platforms accounted.

**Comment by TurkeyMan** [comment]
I changed the names to Gcc/Clang and populated all targets.

**Comment by TurkeyMan** [comment]
This one is blocking all `std::`, the others are blockers to `vector`/`string`. Can this PR has priority?

**Review by Geod24 — APPROVED** [review]
LGTM

**Review by jacob-carlborg — APPROVED** [review]

**Comment by jacob-carlborg** [comment]
Since this is a new feature I think this should have an explicit changelog entry. Feel free to add that in a follow up PR.

File: src/dmd/mars.d (hunk 3 above)
Comment (ibuclaw) [reply src/dmd/mars.d]: >  That's how it's done for the C standard library: https://github.com/dlang/druntime/blob/master/src/core/stdc/locale.d#L60-L137.

Speaking of which, why don't we just add `CRuntime_OSX` and `CRuntime_BSD`?

I take it that FreeBSD, DragonFlyBSD, OpenBSD, ..., InsertNameHereBSD all share the same libc away.

**Comment by RazvanN7** [comment]
Resolve the conflict and we're good to go!

**Comment by thewilsonator** [comment]
@jacob-carlborg this still has a merge conflict...

File: src/dmd/mars.d [hunk 4]
```diff
@@ -1307,18 +1307,30 @@ void addDefaultVersionIdentifiers(const ref Param params)
...
+            VersionCondition.addPredefinedGlobalIdent("CppRuntime_DigitalMars");
+        }
     }
     else if (params.isLinux)
     {
         VersionCondition.addPredefinedGlobalIdent("Posix");
         VersionCondition.addPredefinedGlobalIdent("linux");
         VersionCondition.addPredefinedGlobalIdent("ELFv1");
+        VersionCondition.addPredefinedGlobalIdent("CRuntime_Glibc");
+        VersionCondition.addPredefinedGlobalIdent("CppRuntime_Gcc");
```
[the remark this answers is not in the document]
Comment (ibuclaw) [reply src/dmd/mars.d]: The C standard library is a standalone library, in which you can mix any combination of compiler and runtime - on Linux you also have `CRuntime_Musl`, `CRuntime_UClibc`, and `CRuntime_Bionic`.

The C++ standard library is tied to, and shipped with the compiler.

File: src/dmd/mars.d [hunk 5]
```diff
@@ -1307,18 +1307,30 @@ void addDefaultVersionIdentifiers(const ref Param params)
...
     }
     else if (params.isOSX)
     {
         VersionCondition.addPredefinedGlobalIdent("Posix");
         VersionCondition.addPredefinedGlobalIdent("OSX");
-
+        VersionCondition.addPredefinedGlobalIdent("CppRuntime_Clang");
```
[the remark this answers is not in the document]
Comment (TurkeyMan) [reply src/dmd/mars.d]: I wondered that myself.

File: src/dmd/mars.d [hunk 6]
```diff
@@ -1307,18 +1307,30 @@ void addDefaultVersionIdentifiers(const ref Param params)
     if (params.isWindows)
     {
         VersionCondition.addPredefinedGlobalIdent("Windows");
+        if (global.params.mscoff)
+        {
+            VersionCondition.addPredefinedGlobalIdent("CRuntime_Microsoft");
+            VersionCondition.addPredefinedGlobalIdent("CppRuntime_Microsoft");
```
[the remark this answers is not in the document]
Comment (TurkeyMan) [reply src/dmd/mars.d]: I initially did use the CRuntime to imply the C++ runtime, but I added the CppRuntime in response to feedback on the `std::array` PR. The reason was that while Windows has a strong connection, other platforms do not.

MSVC C and C++ runtimes are tightly coupled, and glibc typically implies libstdc++, but C runtimes eg; bionic, musl, uclibc, all don't infer a C++ runtime. Android offers 2 or 3 C++ runtimes to select from, and libc++ is relatively new and not mandated by LLVM toolchains.

Embedded toolchains are likely to use some combination of embedded libs; newlib? stlport? etc... I was persuaded by the opinion that these should be separated.

File: src/dmd/mars.d (hunk 5 above)
[the remark this answers is not in the document]
Comment (jacob-carlborg) [reply src/dmd/mars.d]: Clang does not provide a C runtime, the platform does.

[the remark this answers is not in the document]
Comment (TurkeyMan) [reply src/dmd/mars.d]: I think he might have rather meant:

> Why is there no CRuntime_AnyRuntimeAtAll ?

At least, that's what I was thinking ;)
There just isn't one defined for most of those platforms for some reason.

**Pull request merged**